### PR TITLE
Decouple the notebook from the widget manager.

### PR DESCRIPTION
**Pull request by jdfreder**
Separate the parts of the widget manager than need comm_manager and notebook instances to work from the base.  Allows people to write there own widget managers, without a notebook or comm_manager.

### CODE REVIEW

**Comment by jdfreder** [comment]
@rgbkrk is interested in this

**Comment by rgbkrk** [comment]
Sweet, thanks for the ping. I'll watch this closely, use it in the notebook, and review where I can.

**Comment by rgbkrk** [comment]
From Jon in chat:

> We'll also need to make sure `ipywidgets` can be `npm` installed, in a follow on PR.
> In case you want to experiment while I'm gone, you'll need to inherit from the manager base to create your own manager then, there are a couple methods that need to be overridden
> 
> I'm saying all this in case you want to experiment next week while I'm gone

**Comment by jdfreder** [comment]
@rgbkrk @SylvainCorlay this is ready for review : )

**Comment by rgbkrk** [comment]
`/me drools`

File: ipywidgets/static/widgets/js/manager.js [hunk 1]
```diff
@@ -461,54 +335,248 @@ define([
...
+                    return that.display_model(undefined, model, options);
                 }));
-            });
+            }));
         }).catch(utils.reject('Could not set widget manager state.', true));
     };
 
-    WidgetManager.prototype._get_connected_kernel = function() {
+    ManagerBase.prototype._create_comm = function(comm_target_name, model_id, metadata) {
+        throw new Error("Manager ._create_comm not implemented");
```
Comment (rgbkrk) [inline ipywidgets/static/widgets/js/manager.js]: I'm assuming you didn't mean to put a space after `Manager` and before the `.` here or below.

File: package.json [hunk 2]
```diff
@@ -15,7 +15,7 @@
     "css": "node build_css.js",
     "prebuildtests": "rimraf ./ipywidgets/tests/bin/tests",
     "buildtests": "tsc ./ipywidgets/tests/**/*.ts --outDir ./ipywidgets/tests/bin -d -m commonjs -t ES5",
-    "postinstall": "npm run bower"
+    "prepublish": "npm run bower && npm run build"
```
Comment (rgbkrk) [inline package.json]: :+1:

**Comment by rgbkrk** [comment]
Is it too much to ask that since you created new classes and functions that you could write (at the very least) some small `mocha` tests on the side to run through `node`?

**Comment by jdfreder** [comment]
@rgbkrk I just tried, and mocha tests in this PR are a no-go because the widgets use AMD.  I do not object to node-ifying the widgets in another PR and then adding mocha tests to that.

File: ipywidgets/static/widgets/js/manager.js [hunk 3]
```diff
@@ -295,8 +158,8 @@ define([
     /**
      * Deprecated, use `new_widget` instead.
      */
-    WidgetManager.prototype.create_model = function (options) {
-        console.warn('WidgetManager.create_model is deprecated. Use WidgetManager.new_widget');
+    ManagerBase.prototype.create_model = function (options) {
+        console.warn('ManagerBase.create_model is deprecated. Use ManagerBase.new_widget');
```
Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/manager.js]: Could we move this to WidgetManager, so that we don't carry deprecated classes to the cleaner future?

**Comment by SylvainCorlay** [comment]
A big :+1:  to this. @ChakriCherukuri could be interested to look at this as well.

File: ipywidgets/static/widgets/js/manager.js (hunk 3 above)
Comment (jdfreder) [inline ipywidgets/static/widgets/js/manager.js]: :heavy_check_mark:

**Comment by JanSchulz** [comment]
I was looking into what needs to be done to build a shiny clone (like a notebook, which is executed in the backend and only the output is sent to the frontend -> static content and widgets; only widgets can communicate with the backend via the widget protocol) and one thing I found that the widgets have / need direct access to the kernel. IMO this would prevent such a usage, as this means that a potential public interface can (via the browser js console) submit python code to execute on the kernel.

Is it planned to also decouple the kernel connection from the widgets in this PR?

**Comment by jdfreder** [comment]
Hi @JanSchulz ,

This PR decouples the notebook, kernel, etc from the WidgetManager.   My next step, in another PR, is to do the same for the Widget base class.  Once that's done, all that is required to display widgets outside of the notebook is an implementation of the ManagerBase class and a dummy comm.

I did this work a couple of times in the past, never to be merged.  You can see the last go, almost a year and a half ago [here](https://github.com/ipython/ipython/pull/5900).

**Comment by SylvainCorlay** [comment]
Currently running on this branch and things seem to work.

**Comment by jdfreder** [comment]
Awesome, well it's ready to merge when you are.  I'd be happy to continue my decoupling in more PRs, to follow.

**Comment by rgbkrk** [comment]
Awesome! Once this is shipped if there's more decoupling, I'll try it out in jupyter-sidecar.

/cc @parente

**Comment by lbustelo** [comment]
Basic question... If you take the Comm out of the widget manager... what does it do then? From the perspective of [jupyter-incubator/declarativewidgets](https://github.com/jupyter-incubator/declarativewidgets) we've identified WidgetManager as a class that is core to interactive widgets. What are the use case that is requiring both Comm and Notebook to be decoupled?

**Comment by rgbkrk** [comment]
> What are the use case that is requiring both Comm and Notebook to be decoupled?

* [Thebe](https://github.com/oreillymedia/thebe) isn't really the notebook and should work with only kernel access
* Dashboards could work without a notebook setup
* Dashboards that aren't tied to a live backend
* Atom packages that would like to feature the widgets as well [Hydrogen](https://github.com/willwhitney/hydrogen), [Atom-Notebook](https://github.com/jupyter/atom-notebook)
* Electron apps that are *not* the notebook, e.g. [sidecar](https://github.com/nteract/jupyter-sidecar)

**Comment by lbustelo** [comment]
Thanks @rgbkrk! I think many of these PRs need to start with some documentation of WHY. 

It is hard to keep up and almost feel like unnecessary effort until the dust settles.

**Comment by jdfreder** [comment]
@lbustelo sorry, in the next PR I will try to better explain the "why".  In the following PRs I'll embrace duck typing for the comm.

**Comment by rgbkrk** [comment]
Thank you @lbustelo for asking! I definitely don't want to hide any of this away.

**Comment by SylvainCorlay** [comment]
hop!

**Comment by rgbkrk** [comment]
:tada:

**Comment by SylvainCorlay** [comment]
Can't wait to be able to use pythreejs and bqplot in other front-ends like the sidecar.

**Pull request merged**